### PR TITLE
Anonymous user language selection

### LINKED ISSUE #1295: Allow anonymous users to change languages
Currently a user must create / log in to an account before they can set a language preference for PeeringDB by updating their user profile.  We should allow anonymous users to change the language on any page without having to get through the account login process.

I think it would be best to implement this as a language select box in the header or footer of the site, which will set a language cookie for the user and then reload the page to the selected language.

This is a project I've been discussing with @leovegoda for a while now, but I don't think we have an open Issue for it yet.  I have some code to contribute for this which needs feedback, which I'll include here shortly as a PR with some screenshots.

### LINKED ISSUE #1296: CSRF cookie not set error from email confirmation view
Users without an existing session cookie who press the button to confirm their email will receive an error response:

`{"non_field_errors": ["CSRF cookie not set."]}`

The trigger for this issue seems to be the combination of https://github.com/peeringdb/peeringdb/issues/1205 and https://github.com/peeringdb/peeringdb/pull/1183.  Because we use CSRF middleware with `CSRF_USE_SESSIONS`, we need to have a session for all paths that may take a POST request from an anonymous user.  We need to update the custom session middleware to allowlist all new paths where an anonymous user may make a POST request.

Why this doesn't happen every time: the email confirmation page does not set a session cookie, but it can use one that already exists from the user having visited one of the login pages allowlisted in our custom Session Middleware: https://github.com/peeringdb/peeringdb/blob/355fff42924a62b1920bd1f263d83e696e96b74b/peeringdb_server/middleware.py#L52-L58.  Note that this does not convey any additional security - the user could simply visit the `/login` page to receive a new session cookie, without actually needing to login.

I suggest that we invert the logic in the session middleware to instead add the API endpoints to a denylist, and then allow session creation for all other paths.  This will make it simpler to add new paths with CSRF protection in the future.

For example, I would like to add a language selection option for anonymous users on all non-API pages in https://github.com/peeringdb/peeringdb/issues/1295.

**To Reproduce**
1. Create a new user profile.
1. Delete your session cookie.
1. Open the confirmation email link.
1. Click the confirm button.
1. Failure.

### PR DESCRIPTION
**Pull request by dvanallen**
Language selection in header:

Screenshot 2023-01-02 at 3 40 36 AM

Language selection with different language, showing submit button hover style:

Screenshot 2023-01-02 at 3 40 55 AM

Fixes #1295, Fixes #1296

### CODE REVIEW

File: peeringdb_server/middleware.py [hunk 1]
```diff
@@ -46,28 +45,20 @@ def process_response(self, request, response):
...
 
-            NEW_SESSION_VALID_PATHS = [
-                reverse("login"),
-                reverse("register"),
-                reverse("username-retrieve"),
-                reverse("username-retrieve-initiate"),
-                reverse("reset-password"),
+            no_session_prefixes = [
+                "api/",
+                "api-auth/",
```
Comment (grizz) [inline peeringdb_server/middleware.py:52]: Allowing everything except these has performance implications @peeringdb/oc will have to look at

**Comment by grizz** [comment]
Waiting per ops discussion

**Comment by ccaputo** [comment]
Saw the comment: `# allow session creation for all request paths except the API`

We definitely can't have a session created on the server side for every unique connection.

I think the work around for this would be to have a client-side cookie, with no matching server state, which saves the language preference.

**Comment by dvanallen** [comment]
>I think the work around for this would be to have a client-side cookie, with no matching server state, which saves the language preference.

These commits do not save the language preference to the user's session on the server side.  I do activate the translation for the current server thread, but that doesn't require a session.  The only requirement to have a session is driven by the fact that we want to make a POST request to set the language preference (the [Django i18n view](https://github.com/django/django/blob/fd21f82aa82b0d75a161f618ef944ebe0923e0ab/django/views/i18n.py#L25) mentions that this should not be done with a GET request), and PeeringDB uses CSRF Middleware with `CSRF_USE_SESSIONS` enabled.  That means we're required to have a session containing a CSRF token.  Or we could disable CSRF protection for this anonymous language endpoint (but note that this would make an eventual migration to the built-in set_language view more difficult).

>We definitely can't have a session created on the server side for every unique connection.

Can you provide any details about the performance issues that were seen previously?  I was hoping that excluding the API endpoints in the session middleware would go a long way towards reducing the amount of sessions being created.  Do we have some idea of which paths were leading to the excessive session creation?

I understand that there is another effort to migrate the PeeringDB session backend to redis, so if the performance concerns are blocking this change today, then let's revisit once we have the new backend.

**Comment by ccaputo** [comment]
> > We definitely can't have a session created on the server side for every unique connection.
> 
> Can you provide any details about the performance issues that were seen previously? I was hoping that excluding the API endpoints in the session middleware would go a long way towards reducing the amount of sessions being created. Do we have some idea of which paths were leading to the excessive session creation?
> 
> I understand that there is another effort to migrate the PeeringDB session backend to redis, so if the performance concerns are blocking this change today, then let's revisit once we have the new backend.

IIRC, when we previously created a session for all non-API requests, a cache table quickly ran into the millions, resulting in AWS RDS slowness for that table - even with an expected index, in addition to increased costs. Restricting sessions to logged in users fixed the issue.

**Comment by grizz** [comment]
I owe the ops mailing list a thread with this info, thanks for dumping in here!

In short, I can think of a few options:

- Move sessions off of RDS (which I'm already looking at)
- Remove sessions period (jwt)
- Move CSRF token back to a cookie

For this specifically, we probably want to wait for Django 4, which we have scheduled for this year, as @dvanallen pointed out that they are changing the POST required behavior in favor of a client side cookie in that https://github.com/django/django/pull/15902

**Comment by dvanallen** [comment]
>For this specifically, we probably want to wait for Django 4, which we have scheduled for this year, as @dvanallen pointed out that they are changing the POST required behavior in favor of a client side cookie in that https://github.com/django/django/pull/15902

The Django change I mentioned to you in Slack is https://github.com/django/django/pull/10875 (I think there may have been an Issue # 15902 at one time which was referenced).

In that PR, the built-in `set_language` endpoint was changed to not store the language preference in the user's session.  However **there was no change to the requirement to use a POST request**.  That is the same before and after the change.  The endpoint I added in this PR is the same in that regard, because it has the same effect of changing how the user will see the rest of the site.

**Comment by grizz** [comment]
@dvanallen Yeah, it refs back to the original issue which is similar to our use case.

Disclaimer: I'm not an expert here, but I have dug into this a bit now.

I don't think you need to use POST to set the active language, only to use the `set_language` redirect.

From reading here:
https://stackoverflow.com/a/2606826
https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#explicitly-setting-the-active-language

I think you can set the language via just setting the cookie (so best here would be via js I think?).

From the way I read https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#the-set-language-redirect-view `set_language` needs to be a POST since it's setting the cookie and doing a redirect https://github.com/django/django/blob/4470c2405c8dbb529501f9d78753e2aa4e9653a2/django/views/i18n.py#L46, but you can just set the cookie and reload the page without using that, right?

That said, I clearly haven't tried it to see if it works either. :)

**Comment by dvanallen** [comment]
Thanks for the feedback all - given the current constraints, I have put together another PR #1319 to make this change using Javascript to set the cookie as suggested.

**Pull request closed without merging**